functional autograd api

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34066 functional autograd api**

Basic implementation of https://github.com/pytorch/pytorch/issues/30632

Differential Revision: [D20260307](https://our.internmc.facebook.com/intern/diff/D20260307)

albanD committed:
functional autograd api

[ghstack-poisoned]

albanD committed:
Update on "functional autograd api"




[ghstack-poisoned]

albanD committed:
Update on "functional autograd api"




[ghstack-poisoned]

albanD committed:
Update on "functional autograd api"


Basic implementation of https://github.com/pytorch/pytorch/issues/30632

[ghstack-poisoned]

albanD committed:
Update on "functional autograd api"


Basic implementation of https://github.com/pytorch/pytorch/issues/30632

[ghstack-poisoned]

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,714 @@
+def _tuple_postprocess(inp, tuple_inp):
+    # Unpacks a potentially nested tuple of Tensors
+    # It is used to invert _as_tuple before returning to the user
+    if isinstance(tuple_inp, tuple):
```
So `tuple_inp` can be either a boolean or a Pair of boolean? I'm not really sure what it does when it is a tuple.

albanD replied:
When its a tuple of two values, it applies the logic twice in a nested manner.
Inp is always a tuple of tuple of Tensors in that case, and tuple_inp[0] is used to remove the first level of tuple in inp. And tuple_inp[1] is used to remove the inner tuple of inp.

This is used for Jacobian and hessian.

gchanan replied:
this PR looks pretty strange -- the first return of as_tuple is a boolean called tuple_inp, which is the same as the argument to _tuple_postprocess, but that value is never used as a bool.

albanD replied:
It is used as a bool for functions that return something that match the shape of a single thing (input or output of the user function).
For results that match the shape of two things (like output + input for jacobian), we could do two calls to `_tuple_postprocess` for each boolean tuple_*, but I though it was clearer to combine them in a tuple and have the function handle the nesting itself.

Is that the answer you're looking for? Or maybe I am not understanding the question?

zou3519 replied:
Just pulling on this thread a little: it's confusing that:
- `tuple_inp` is used here and in the previous function (as_tuple), but the tuple_inp here can be different from in as_tuple. The confusing part is that `_tuple_postprocess` is advertised as the inverse of `tuple_inp` (it is not always the inverse, it does something slightly different when `tuple_inp` is a tuple).
- `tuple_inp` has the word "tuple" in it. This makes it hard for the reader to reason about what its type is -- "is it a bool? is it a tuple"? Then the reader realizes that it's either a Tuple[Bool] or a Bool.

It might be clearer to split `_tuple_postprocess` into (1) a function that inverts `_as_tuple` once and (2) some function that undoes the nested _as_tuple, as needed for jacobian and hessian

albanD replied:
Will the following changes work:
- change `tuple_inp` to `is_inp_tuple`
- advertise `_as_tuple` as ensuring everything is a tuple of Tensor.
- advertise `_tuple_postprocess` as doing unpacking of tuples to match the user's original expectations.

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,714 @@
+def _grad_preprocess(inputs, create_graph, need_graph):
+    # Preprocess the inputs to make sure they require gradient
+    # inputs is a tuple of Tensor to preprocess
+    # create_graph specifies if the user wants gradients to flow back to the original input
```
"original input" is not the same as the `inputs` arg to this function, right?

albanD replied:
Yes, original input is `inputs` here. Will change.

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,714 @@
+    # Preprocess the inputs to make sure they require gradient
+    # inputs is a tuple of Tensor to preprocess
+    # create_graph specifies if the user wants gradients to flow back to the original input
+    # need_graph specifies if we internally want gradients to flow back for this Tensor
```
Which Tensor? `inputs`?

albanD replied:
For the corresponding entry that is in `res`. I'll make that clearer in the comment.

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,714 @@
+
+def jvp(func, inputs, v=None, create_graph=False, strict=False):
+    r"""Function that computes the dot product between  the Jacobian of
+    the given function at the point given by the inputs and a vector ``v``.
```
Was the conclusion that we're not going to note in the documentation that this isn't too efficient b/c it is doing the double grad trick?

albanD replied:
The thing is that it is the most efficient thing we can do right now. So I did not mentionned it. If you think we should, I can add it.

btw for the hvp, I did add a note that there is a more efficient version that computes the same thing in some cases.

zou3519 replied:
I'm not sure how to answer the question.

On the one hand, not mentioning means that users won't be scared away from using `jvp`. On the other hand, if we don't mention it, some power users might come after us and do some pytorch vs jax benchmarking comparisons. Those aren't the worse thing in the world though because it is true that today this is the best we can support

zou3519 replied:
One alternative could be that we quote that we "compute the jvp via the double backward trick"

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,714 @@
+    for inp in inputs:
+        if create_graph and inp.requires_grad:
+            # Use .view_as() to get a new Tensor in a differentiable way
+            res.append(inp.view_as(inp))
```
So if `create_graph=True`, we allow the inputs to `vjp` (and the others) to be non-leaf nodes that requires_grad?

albanD replied:
Yes.
After investigation, enforcing them to be leafs does not give any benefit.
In particular, we cannot do `.detach().requires_grad_()` as the user's Tensor won't get gradients otherwise. So we need to do this exact same logic anyway. Since it works in both cases, I removed the constraint that it has to be a leaf.

zou3519 commented on test/test_autograd.py:
```diff
@@ -4129,6 +4130,863 @@ def fn(*inputs):
 
+class TestAutogradFunctional(TestCase):
+    def _assert_same_struct(self, res, base):
+        # base and res should be Tensors or tuple of Tensors containing the same thing
```
Not the same thing, but rather, have the same sizes?

albanD replied:
Yes, that's a typo. Will change.

zou3519 commented on test/test_autograd.py:
```diff
@@ -4129,6 +4130,863 @@ def fn(*inputs):
+        # base and res should be Tensors or tuple of Tensors containing the same thing
+        if torch.is_tensor(base):
+            self.assertTrue(torch.is_tensor(res))
+            self.assertTrue(base.size() == res.size())
```
Does `self.assertEqual(base.size(), res.size())` work? The error message for assertTrue is not the best (it doesn't help when debugging problems in the CI because it just prints "Assertion error: expected True but got False")

albanD replied:
I can try it. I'll let you know if it works.

albanD replied:
Seems to work !

zou3519 commented on test/test_autograd.py:
```diff
@@ -4129,6 +4130,863 @@ def fn(*inputs):
+        self.assertIsNone(vjp_val[0].grad_fn)
+        self.assertIsNone(vjp_val[1].grad_fn)
+
+    def test_vjp_scalar(self):
```
Is there a `vjp(func, input, v=None)` test somewhere?

albanD replied:
There are checks to make sure that v is passed but I don't this this one exist.
I will update this test to `v=None` (and do the same for jvp, vhp and hvp.

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,714 @@
+            ``func``'s output contains a single element and it will be set as a Tensor
+            containing a single ``1``.
+        create_graph (bool, optional): If ``True``, both the output and result will be
+            computed in a differentiable way. Note that when strict if ``False``, the result can not
```
"when strict if" -> "when strict is"

This should be edited in the other docstrings as well

albanD replied:
good catch, will do.

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,714 @@
+    the given function at the point given by the inputs.
+
+    Args:
+        func (function): a Python function that takes Tensor inputs and returns
```
Do we tell the users to capture via lambda things that aren't Tensors?

albanD replied:
Do we want to put that in the docstring of each function? Or in the introduction to the section of the doc that presents all these functional API?
This intro could also mention the subtle difference between this API and the old one.

zou3519 replied:
> Or in the introduction to the section of the doc that presents all these functional API?

Introduction sounds like the better place, otherwise there will be a lot of copy-pasting

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,714 @@
+    Args:
+        func (function): a Python function that takes Tensor inputs and returns
+            a tuple of Tensor or a Tensor.
+        inputs (tuple of Tensor or Tensor): inputs to the function.
```
nit: "inputs to the function `func`" makes it less ambiguous

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,714 @@
+        inputs (tuple of Tensor or Tensor): inputs to the function.
+        v (tuple of Tensor or Tensor): The vector that will multiply the Jacobian. Must be the
+            same size as the output of ``func``. This argument is optional when
+            ``func``'s output contains a single element and it will be set as a Tensor
```
nit: "This argument is optional when ``func``'s output contains a single element and (if it is not provided) it will be set as a Tensor"

Things in the parenthesis to make it unambiguous

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,714 @@
+            ``func``'s output contains a single element and it will be set as a Tensor
+            containing a single ``1``.
+        create_graph (bool, optional): If ``True``, both the output and result will be
+            computed in a differentiable way. Note that when strict if ``False``, the result can not
```
Er what happens if strict is `False` and the result doesn't require gradients? Do we throw an error?

albanD replied:
No, we just create a Tensor full of `0` of the appropriate size that does not require gradients.

When `strict=False`, we never raise an error.

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,714 @@
+            computed in a differentiable way. Note that when strict if ``False``, the result can not
+            require gradients or be disconnected from the inputs.
+            Defaults to ``False``.
+        strict (bool, optional): If ``True``, an error will be raised when we detect we detect that some quantities
```
"we detect we detect" -> "we detect", this is also present in the other docstrings

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,714 @@
+            Defaults to ``False``.
+        strict (bool, optional): If ``True``, an error will be raised when we detect we detect that some quantities
+            are independent of the element we want to differentiate it with. If ``False``, such quantity
+            will only be set to ``0`` which is the expected mathematical value.
```
this "such quantity" (meaning the gradient) is different from the previous "such quantity" (meaning one of the outputs), right? That is a little confusing and it might be better to be more explicit about which quantities we're talking about.

Here's what I understand what the documentation to mean:
If True, an error will be raised when we detect that there exists an input such that all the outputs are independent of it. If False, we return zeros as the vjp for said inputs, which is the expected mathematical value.

albanD replied:
Right, I tried to remain to general. I'll update all the docs.

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,714 @@
+
+    Returns:
+        output (tuple of Tensor or Tensor): output of ``func(inputs)``
+        result (tuple of Tensor or Tensor): result of the dot product with the same shape
```
It feels strange that we're using "shape" or "size" to refer to the "shape of a container and its tensor constituents", but I don't see a better way to describe it

albanD replied:
Yes, we use shape to refer to the nested structure of optional tuple + Tensor. Maybe I should use a word that does not have connotation as being a Tensor size/shape? like structure?

zou3519 replied:
I don't know what the right answer for this is. I'm fine with keeping it as "shape" for now, and then perhaps elaborating on it in the future or picking a new way to describe this.

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,714 @@
+    outputs = func(*inputs)
+    tuple_outputs, outputs = _as_tuple(outputs, "outputs of the user-provided function", "jvp")
+    _check_requires_grad(outputs, "outputs", strict=strict)
+    grad_outputs = tuple(torch.zeros_like(out, requires_grad=True) for out in outputs)
```
Does it matter what the grad_outputs is for the double backward trick?

albanD replied:
No, it's a linear function, so the backward is independent on the value (if there is no nan or inf). I do want to set it to a valid value though to avoid NaNs that could appear with `torch.empty()`. Do you think I should add a comment?

zou3519 replied:
A comment to a resource about the double backward trick would probably be helpful for anyone looking at this code in the future

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,714 @@
+
+        jac_i = tuple([] for _ in range(len(inputs)))
+        for j in range(out.nelement()):
+            # Do this instead of the grad_output trick in `gradcheck` to avoid version counter issues
```
What were the version counter issues?

albanD replied:
When we have `create_graph=True`, and `grad_output` would be saved when creating the graph in the backward, we cannot use a similar trick where we set a `1` at one coordinate one by one. Because the saved `grad_output` in the graph will then have an invalid version.

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,714 @@
+            vj = _autograd_grad(out.reshape(-1)[j], inputs, retain_graph=True, create_graph=create_graph)
+
+            for el_idx, (jac_i_el, vj_el, inp_el) in enumerate(zip(jac_i, vj, inputs)):
+                # Note that grad_out_view might share memory with vj
```
What is "grad_out_view"?

albanD replied:
Ho that's an old comment from when I was using the `grad_ouptut` trick mentionned above (in the version counter note).
This comment, and the `.clone()` on `vj_el` below can be removed. Good catch.

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,714 @@
+                    if strict and create_graph and not vj_el.requires_grad:
+                        raise RuntimeError("The jacobian of the user-provided function is independent of "
+                                           "input {}. This is not allowed in strict mode when create_graph=True.".format(i))
+                    jac_i_el.append(vj_el.clone())
```
Is it sufficient to just detach? `jac_i_el.append(vj_el.detach())`?

albanD replied:
No you don't want to always detach() here. That logic is in `_grad_postprocess`.
As mentioned above, this is a remain of a previous iteration, the `.clone()` can just be removed here.

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,714 @@
+        return jac
+
+    res = jacobian(jac_func, inputs, create_graph=create_graph, strict=strict)
+    # False here to remove the tuple dimension from the output as we know it does not have any.
```
Where is the `False`?

albanD commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,714 @@
+            linearized inputs and output. If one of the two is a tuple, then the Jacobian
+            will be a tuple of Tensors. If both of them are tuples, then the Jacobian will
+            be a tuple of tuple of Tensors where ``Jacobian[i][j]`` will contain the Jacobian
+            of the ``i``th output and ``j``th input and will have as size the sum of the size
```
Adding is well define in the sense that `a.size() + b.size()` does what we want here.
But I can swap the doc to concatenate as it sounds better indeed.

albanD committed:
Update on "functional autograd api"


Basic implementation of https://github.com/pytorch/pytorch/issues/30632

[ghstack-poisoned]

zou3519 commented on test/test_autograd.py:
```diff
@@ -4129,6 +4130,877 @@ def fn(*inputs):
+        with self.assertRaisesRegex(RuntimeError, "v is a tuple of invalid length: should be 1 but got 2."):
+            res = autogradF.vjp(foo, inp, (torch.ones_like(inp), torch.ones_like(inp)))
+
+        with self.assertRaisesRegex(RuntimeError, "Entry 0 in v has invalid size: should be torch.Size"):
```
Hmm, "Entry 0 in v" is misleading when the user passes is a Tensor (not a Tuple) for v (`v = torch.ones(2)` in this case). Sometimes users will take this to mean "v[0]"

albanD replied:
Special casing all the error messages for this is going to be quite verbose, and might not help with code clarity.
I can add to the preamble that single Tensors are considered the same as a tuple with a single element that is the given Tensor?

zou3519 replied:
This is a pretty small point so I'm fine with your proposal of adding something to the preamble or just worrying about this later.

Here are some alternative ways I brainstormed to handle this:
The other cases have error messages that read as "input 2 ..." or "output 0 ..." which is clearer than "Entry 0 in v" ("entry" is ambiguous -- does it mean v[0] or v?) Maybe an error message like "The 0th v" would be better, but then we'd need logic to conjugate the number "0th, 1st, 2nd, ..., 11th" and that can be awkward. As an alternative, it's not too hard to just special case the 'v' error message - we could just pass a `was_v_a_tensor` to `_validate_v`.

albanD replied:
Updated using a boolean flag.

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,712 @@
+    return _tuple_postprocess(outputs, tuple_outputs), _tuple_postprocess(jvp, tuple_outputs)
+
+
+def jacobian(func, inputs, create_graph=False, strict=False):
```
Hmm, I noticed that jax has jacfwd and jacrev. Do you have a sense of, if we do need and implement efficient forward mode AD, if we should have a split in the API like that?

albanD replied:
jax also has a `jax.jacobian` function. I think given that we don't have a forward mode, the general function is enough. If we add forward mode later, we can do like them and have all 3 functions (I think jax.jacobian is just the same as jax.jacrev)

zou3519 commented on test/test_autograd.py:
```diff
@@ -4129,6 +4130,877 @@ def fn(*inputs):
+
+        inp = torch.rand(4)
+        with self.assertRaisesRegex(TypeError, "The inputs given to vjp must be either a Tensor"):
+            res = autogradF.vjp(foo, (inp, 2))
```
I can see someone "reversing" the v-check and the inputs check in the future. Should we pass in a correct `v` here to make it clear that we are testing incorrect inputs?

albanD replied:
Good point. Will do !

zou3519 replied:
Err actually, you can't "reverse" the checks because we need to check that the shape of `v` matches the output. The other point still stands, though, it might be nice to pass in a valid v here to make it clear what we're testing

albanD replied:
True for vjp. But not for jvp and hvp. I should at least update these two.

zou3519 commented on test/test_autograd.py:
```diff
@@ -4129,6 +4130,877 @@ def fn(*inputs):
+        res = autogradF.vjp(foo, inp, torch.ones(3))[1]
+        self._assert_same_struct(res, inp)
+
+    def test_vjp_err_check_strict(self):
```
I don't see a test for "The jacobian of the user-provided function is independent of..." (the case where we are strict, have create_graph=True, but the vjp doesn't require grad)

albanD replied:
Good catch, this error was added at the end and I forgot to update the tests. Will do

zou3519 commented on test/test_autograd.py:
```diff
@@ -4129,6 +4130,877 @@ def fn(*inputs):
+        res = autogradF.vjp(reducer, inputs, v)
+        self._assert_same_struct(res[1], inputs)
+        self.assertIsNone(res[0].grad_fn)
+        self.assertIsNone(res[1].grad_fn)
```
Is this equivalent to checking if the tensor requires grad?

albanD replied:
Not in general. But here very close.
The main difference being that this won't catch leafs that require gradients. Only Tensors that are "inside" the graph (not leafs).

zou3519 replied:
Should we switch these to check `self.assertFalse(res[0].requires_grad)` since it also includes the case of leafs that require gradient? (The marginal benefit for adding this doesn't seem like too much because one really should not be getting leaves that require grad out of `vjp` so I think either way works)

zou3519 commented on test/test_autograd.py:
```diff
@@ -4129,6 +4130,877 @@ def fn(*inputs):
+        self._assert_same_struct(res[0], v)
+        self._assert_same_struct(res[1], inputs)
+
+    def test_vjp_create_graph(self):
```
Should we also test that the `create_graph` behavior  "extends" the graph if the inputs to `vjp` are not leaf nodes? i.e., from what I understand, the following should work:

```
x = torch.randn(4, requires_grad=True)
y = x.cos()
_, vjp = autogradF.vjp(lambda x: x ** 2, y, ones_like(y), create_graph=True)
result = autograd.grad(vjp, x, ones_like(vjp))
````

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,712 @@
+            require gradients or be disconnected from the inputs.
+            Defaults to ``False``.
+        strict (bool, optional): If ``True``, an error will be raised when we detect that there exists an input
+            such that all the outputs are independent of it. If ``False``, we return zeros as the vjp for
```
"vjp" -> "jvp"

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,712 @@
+            require gradients or be disconnected from the inputs.
+            Defaults to ``False``.
+        strict (bool, optional): If ``True``, an error will be raised when we detect that there exists an input
+            such that all the outputs are independent of it. If ``False``, we return zeros as the vjp for
```
"vjp" -> "jacobian"? The copy-paste here for the other docstrings might be wrong too

albanD replied:
Yep, bad copy paste...

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,712 @@
+        >>> def pow_reducer(x):
+        ...   return x.pow(3).sum()
+        >>> inputs = torch.rand(2, 2)
+        >>> hessian(pow_reducer, inputs)
```
It might be nicer to send in `inputs` as a tensor with one dimension so that the output comes out as a matrix.

Also, I've noticed that a lot of the examples use random inputs. Would it be better if they didn't use random inputs (and we instead did `inputs = torch.tensor([1., 2.])` for example) so that the user can "check" that the output matches their definition of hessian?

albanD replied:
After chatting with @gchanan I will postpone this to a later PR.

zou3519 commented on test/test_autograd.py:
```diff
@@ -4129,6 +4130,877 @@ def fn(*inputs):
+        err_msg_out = "The Tensor returned by the function given to hessian should contain a single element"
+        with self.assertRaisesRegex(RuntimeError, err_msg_out):
+            res = autogradF.hessian(bar2, inp)
+
```
I don't think there's a test for the "The function given to hessian should return a single Tensor" error message

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,712 @@
+
+    def ensure_single_output_function(*inp):
+        out = func(*inp)
+        tuple_out, t_out = _as_tuple(out, "outputs of the user-provided function", "hessian")
```
nit: We don't really need this check right? We just need to check if `torch.is_tensor(out)`

albanD replied:
The reason for this is:
- We want t_out to be able to use _check_requires_grad
- That gives nice error messages for non-Tensor returns

zou3519 replied:
Oh I see, _as_tuple gives a nice error message here. Thanks for the clarification, I agree with keeping it.

zou3519 commented on test/test_autograd.py:
```diff
@@ -4129,6 +4130,877 @@ def fn(*inputs):
+            return x.sum()
+        inputs = torch.rand(4, 4)
+        res = autogradF.hessian(reducer, inputs)
+        self._assert_interleaved_struct(res, inputs, inputs)
```
This case is sort of redundant with `test_hessian_output` because the output must be scalar

albanD replied:
Yes, I just wanted to re-test all the cases in this function to make sure not to forget any. You think I should remove it?

zou3519 replied:
Nah, redundant tests are better than not being sure if the tests cover all our cases

zou3519 commented on test/test_autograd.py:
```diff
@@ -4129,6 +4130,877 @@ def fn(*inputs):
+        self._assert_same_struct(res, inputs)
+
+        def bad_reducer(x):
+            return x.sum().view(1, 1, 1)
```
nit: x.sum(keepdim=False) does something similar

albanD replied:
Not really because `x` has 2 dimensions, and the view will give 3.

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,712 @@
+                               "is a single Tensor with a single element.")
+
+    outputs = func(*inputs)
+    tuple_outputs, outputs = _as_tuple(outputs, "outputs of the user-provided function", "vhp")
```
Is this necessary? We already check below that `torch.is_tensor(outputs)`

albanD replied:
Yes because we want a tuple of Tensor. All our utility functions assume we work with tuple of Tensor.

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,712 @@
+            if strict and create_graph and not vhp_i.requires_grad:
+                raise RuntimeError("The hessian of the user-provided function is independent of "
+                                   "input {}. This is not allowed in strict mode when create_graph=True.".format(i))
+        vhp += (vhp_i,)
```
This logic, post-processing None gradients from autograd.grad, seems common among all of the functions. Would it be worth it to toss this into a helper function?

The error messages look like they are going to need to be different so this might turn out to be a bad idea if the result of the move makes things less readable

albanD replied:
The error messages are different, but I guess I can use a flag for that.
Will refactor that into a single function 👍

albanD committed:
Update on "functional autograd api"


Basic implementation of https://github.com/pytorch/pytorch/issues/30632

Differential Revision: [D20260307](https://our.internmc.facebook.com/intern/diff/D20260307)

[ghstack-poisoned]

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,712 @@
+            if strict:
+                raise RuntimeError("The hessian of the user-provided function is independent of "
+                                   "entry {} in the grad_jacobian. This is not allowed in strict "
+                                   "mode as it prevents form using the double backward trick to "
```
"prevents form" -> "prevents from"

zou3519 commented on docs/source/autograd.rst:
```diff
@@ -11,6 +11,32 @@ Automatic differentiation package - torch.autograd
+
+This section contains the higher level API for the autograd that build on the basic API above.
+
+This API always work with user-provided functions that take only Tensors as input and return
```
"always work" -> "always works"

zou3519 commented on docs/source/autograd.rst:
```diff
@@ -11,6 +11,32 @@ Automatic differentiation package - torch.autograd
+
+This API always work with user-provided functions that take only Tensors as input and return
+only Tensors.
+If your function takes other arguments that are not Tensors and that should be ignored for the
```
This doesn't tell us what happens if we want other arguments to be Tensors but not compute their derivatives so it's a bit ambiguous. Maybe an example that captures another tensor would help here (or a rephrasing as "other arguments that are not Tensors and should be ignored, or arguments that are Tensors that should not be differentiated")

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,719 @@
+    for inp in inputs:
+        if create_graph and inp.requires_grad:
+            # Use .view_as() to get a new Tensor in a differentiable way
+            res.append(inp.view_as(inp))
```
nit: I wonder if `inp[:]` is a more canonical way to get a new view of the tensor, but the comment does explain what's going on here

albanD replied:
I don't like advanced indexing here because it is an implementation detail that it returns a view. `.view_as()` is guaranteed to return a view.

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,719 @@
+
+
+def _grad_postprocess(inputs, create_graph):
+    # Postprocess the generated Tensors to avoid returning Tensors with history when the user does not
```
nit: "when the user does not requested it" -> "when the user did not request it"

albanD committed:
Update on "functional autograd api"


Basic implementation of https://github.com/pytorch/pytorch/issues/30632

Differential Revision: [D20260307](https://our.internmc.facebook.com/intern/diff/D20260307)

[ghstack-poisoned]

zou3519 commented on docs/source/autograd.rst:
```diff
@@ -11,6 +11,33 @@ Automatic differentiation package - torch.autograd
+Functional higher level API
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This section contains the higher level API for the autograd that build on the basic API above.
```
nit: "that build on the basic api above" -> "that builds on the basic api above"

zou3519 commented on docs/source/autograd.rst:
```diff
@@ -11,6 +11,33 @@ Automatic differentiation package - torch.autograd
+
+This section contains the higher level API for the autograd that build on the basic API above.
+
+This API always works with user-provided functions that take only Tensors as input and return
```
nit: the "always" isn't necessary

albanD committed:
Update on "functional autograd api"


Basic implementation of https://github.com/pytorch/pytorch/issues/30632

Differential Revision: [D20260307](https://our.internmc.facebook.com/intern/diff/D20260307)

[ghstack-poisoned]

gchanan commented on docs/source/autograd.rst:
```diff
@@ -11,6 +11,33 @@ Automatic differentiation package - torch.autograd
+Functional higher level API
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This section contains the higher level API for the autograd that builds on the basic API above.
```
nit: since you jump into discussing the jacobian below, why don't you give some examples of what the "high level" autograd functions are, e.g.

"This section contains the higher level API for the autograd that builds on the basic API above and allows you to compute jacobians, hessians, etc.

gchanan commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,717 @@
+    for inp in inputs:
+        if create_graph and inp.requires_grad:
+            # Use .view_as() to get a new Tensor in a differentiable way
+            res.append(inp.view_as(inp))
```
won't this not work if `inp` is a sparse tensor?

albanD replied:
Good point.
Do we have any operation that would return a view of the same sparse tensor? Or do we have to use `.clone()`?

gchanan replied:
I don't believe we do, unfortunately.  So I think you'd have to use clone for now.

gchanan commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,717 @@
+    # This has the extra constraint that inputs has to be a tuple
+    if torch.is_tensor(outputs):
+        outputs = (outputs,)
+        grad_outputs = (grad_outputs,)
```
do you need to check that outputs and grad_outputs have the same structure?

albanD replied:
Yes I simplified this: we assume we get a tuple of inputs and assert accordingly.

gchanan commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,717 @@
+            elif input_type == "jacobian":
+                raise RuntimeError("The jacobian of the user-provided function with respect to input {}"
+                                   " is independent of the input. This is not allowed in strict mode."
+                                   " You should ensure that your function is twice differentiable and that"
```
maybe a dumb question, but why do I need my function to be twice differentiable to compute the jacobian?

albanD replied:
This is used when we compute the Jacobian as an intermediary result for a hessian. In that case, we need the Jacobian to require gradients.

gchanan replied:
could you make the error message clearer, like "While computing the hessian, found that the Jacobian ..."

albanD replied:
Sure

gchanan commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,717 @@
+
+    Args:
+        func (function): a Python function that takes Tensor inputs and returns
+            a tuple of Tensor or a Tensor.
```
nit: tuple of Tensors (a few places here)

gchanan commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,717 @@
+        func (function): a Python function that takes Tensor inputs and returns
+            a tuple of Tensor or a Tensor.
+        inputs (tuple of Tensor or Tensor): inputs to the function ``func``.
+        v (tuple of Tensor or Tensor): The vector that will multiply the Jacobian. Must be the
```
nit: this seems to suggest you actually multiply the Jacobian, but that isn't true, right?

albanD replied:
Multiply in the matrix sense. I could say "The vector used in the product with the Jacobian" but it doesn't feel much better. You think I should change it?

gchanan replied:
I'd put something like "the vector for which the vector jacobian product is computed."  That way you don't say how it is used, just that a vector jacobian product is computed.

gchanan commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,717 @@
+            require gradients or be disconnected from the inputs.
+            Defaults to ``False``.
+        strict (bool, optional): If ``True``, an error will be raised when we detect that there exists an input
+            such that all the outputs are independent of it. If ``False``, we return zeros as the vjp for
```
nit: return a Tensor of zeros

gchanan commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,717 @@
+            Defaults to ``False``.
+
+    Returns:
+        output (tuple of Tensor or Tensor): output of ``func(inputs)``
```
are these named tuples?  I'd probably call these something like "func_output" and "vjp" or something.

albanD replied:
No they are not named tuples, just regular ones.

gchanan commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,717 @@
+                [2.2046, 1.1292, 1.1432, 1.3059],
+                [1.3225, 1.6652, 1.7753, 2.0152]], grad_fn=<MulBackward0>))
+
+        >>> def adder(x, y):
```
maybe it makes sense to have an example of multiple inputs and multiple outputs?

albanD replied:
My problem is that it is hard to have an example that is big enough so that the dimensions are readable and not just many brackets that are hard to read. While being small enough not to spam the screen too much.
This is why I stopped at multiple inputs only.

gchanan replied:
that sounds fine then.

gchanan commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,717 @@
+        func (function): a Python function that takes Tensor inputs and returns
+            a tuple of Tensor or a Tensor.
+        inputs (tuple of Tensor or Tensor): inputs to the function ``func``.
+        v (tuple of Tensor or Tensor): The vector that will multiply the Jacobian. Must be the
```
same "will multiply" case as above.

gchanan commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,717 @@
+    Note::
+
+        The jvp is currently computed using the double backward trick as we don't have support for
+        forward mode AD in pytorch at the moment.
```
nit: PyTorch

gchanan commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,717 @@
+
+    Note::
+
+        The jvp is currently computed using the double backward trick as we don't have support for
```
is there a reference you can use for the double backward trick?

albanD replied:
I wasn't sure what was the original reference for it. I think this is widely known for people working on autograd.

gchanan replied:
Yes, but more people than just autograd experts will read this documentation.  I think it's fine to just describe it briefly e.g. "compute the backwards of the backward (sometimes called the double backwards trick)"

gchanan commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,717 @@
+            a tuple of Tensor or a Tensor.
+        inputs (tuple of Tensor or Tensor): inputs to the function ``func``.
+        create_graph (bool, optional): If ``True``, the Jacobian will be computed in
+            a differentiable manner. Note that when strict is ``False``, the result can not
```
nit: I think this should be `strict`?

gchanan commented on torch/autograd/functional.py:
```diff
@@ -0,0 +1,717 @@
+
+        This function is significantly slower than `vhp` due to backward mode AD constraints.
+        If your functions is twice continuously differentiable, then hvp = vhp.t(). So if you
+        know that your function verifies this condition, you should use vhp instead that is
```
verifies -> satisfies

albanD committed:
Update on "functional autograd api"


Basic implementation of https://github.com/pytorch/pytorch/issues/30632

Differential Revision: [D20260307](https://our.internmc.facebook.com/intern/diff/D20260307)

[ghstack-poisoned]

albanD committed:
Update on "functional autograd api"


Basic implementation of https://github.com/pytorch/pytorch/issues/30632

Differential Revision: [D20260307](https://our.internmc.facebook.com/intern/diff/D20260307)

[ghstack-poisoned]

gchanan commented on docs/source/autograd.rst:
```diff
@@ -11,6 +11,37 @@ Automatic differentiation package - torch.autograd
+
+This API works with user-provided functions that take only Tensors as input and return
+only Tensors.
+If your function takes other arguments that are not Tensors or Tensors for which you don't require gradients,
```
nit: I'd be explicit here, "for which you don't require gradients" you mean -- the Tensors don't have `requires_grad` set, right?  i.e. I might have requires_grad set but not actually require the gradients.

albanD replied:
Is this better:
"If your function takes other arguments that should be ignored when computing the given quantity (for example inputs that are not Tensors, or Tensors that should be ignored when computing the quantity),"

The example below for the jacobian should give a concrete example.

gchanan replied:
why not just "or Tensors that don't have requires_grad set" ?

albanD committed:
Update on "functional autograd api"


Basic implementation of https://github.com/pytorch/pytorch/issues/30632

Differential Revision: [D20260307](https://our.internmc.facebook.com/intern/diff/D20260307)

[ghstack-poisoned]

zou3519 reviewed (approved):
lgtm as well

albanD committed:
Update on "functional autograd api"


Basic implementation of https://github.com/pytorch/pytorch/issues/30632

Differential Revision: [D20260307](https://our.internmc.facebook.com/intern/diff/D20260307)

[ghstack-poisoned]

gchanan commented on docs/source/autograd.rst:
```diff
@@ -11,6 +11,37 @@ Automatic differentiation package - torch.autograd
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. warning::
+    This API is experimental.
```
nit: I'd put a little more info on what experimental means -- that the performance still needs to be improved and the API might change (although the later is unlikely).

gchanan commented on test/test_autograd.py:
```diff
@@ -4218,6 +4219,988 @@ def fn(*inputs):
+    def _assert_interleaved_struct(self, res, base1, base2):
+        # base1 and base2 can be Tensors or tuples of Tensors.
+        # If they are tuples, res should be a tuple as well.
+        # The indexing works as follow for base1, base2 being
```
nit: as follows
